File: apil_custom/extrusion_log.py
import frappe
from frappe.utils import flt, time_diff_in_seconds


def before_save(doc, method=None):
	"""Live calculations: Total Input, Die Running Time, Rec%, Output/Hr,
	and the linked BOM lookup.

	Moved here from a database-stored Server Script: identical logic, but
	as a real app file it isn't subject to the RestrictedPython sandbox
	(which turned out to cause a hard-to-diagnose Stock Entry valuation
	bug specifically for the After Submit stock-creation step - see
	apil_custom/shift_production_log.py, which now owns that step instead).
	"""
	total_input = sum([(d.billet_weight or 0) for d in doc.billet_charges])
	doc.total_input = total_input

	if doc.die_in and doc.die_out:
		diff = time_diff_in_seconds(doc.die_out, doc.die_in)
		if diff < 0:
			diff += 24 * 3600
		doc.die_running_time = diff

	if doc.output and total_input:
		if doc.output > total_input:
			frappe.throw(
				"Output ({0} kg) cannot exceed Total Input ({1} kg). Please check the entered figures.".format(
					doc.output, total_input
				)
			)
		doc.rec_percent = round((doc.output / total_input) * 100, 2)
	else:
		doc.rec_percent = 0

	if doc.output and doc.die_running_time:
		hours = doc.die_running_time / 3600.0
		doc.output_per_hr = round(doc.output / hours, 2) if hours > 0 else 0
	else:
		doc.output_per_hr = 0

	if doc.sec_no:
		doc.bom = frappe.db.get_value(
			"BOM", {"item": doc.sec_no, "is_active": 1, "is_default": 1, "docstatus": 1}, "name"
		)


def before_submit(doc, method=None):
	"""Block submit (not save) if the warehouse doesn't have enough of the
	BOM's raw material to cover this log's Total Input.

	This is a real-time sanity check for the operator submitting a single
	Die run; it's not the final word - the Shift Production Log that later
	consolidates several logs re-checks the aggregate total when it's
	submitted, since several logs drafted in the same shift can each
	individually pass this check against stock none of them have deducted
	yet (Stock Entries stay Draft until a supervisor submits them).
	"""
	if not doc.bom:
		frappe.throw("No active BOM found for {0}. Cannot validate or create stock movement.".format(doc.sec_no))

	bom_doc = frappe.get_doc("BOM", doc.bom)
	rm_item = bom_doc.items[0].item_code

	available_qty = frappe.db.get_value("Bin", {"item_code": rm_item, "warehouse": doc.source_warehouse}, "actual_qty") or 0

	if flt(doc.total_input) > flt(available_qty):
		frappe.throw(
			"Insufficient {0} stock in {1}: available {2} kg, this log needs {3} kg. "
			"Please arrange more stock (Material Receipt / additional casting) before submitting this log.".format(
				rm_item, doc.source_warehouse, available_qty, doc.total_input
			),
			title="Stock Not Sufficient",
		)


@frappe.whitelist()
def get_extrusion_rm_availability(item_code, warehouse):
	"""Live stock check called from the Extrusion Log form (see
	public/js/extrusion_log.js) as the die/warehouse are picked, so the
	operator sees available billet stock before submitting.

	Migrated from a database-stored Server Script (API type) of the same
	name, kept here as a real whitelisted function instead.
	"""
	result = {"rm_item": None, "available_qty": 0, "bom": None}

	if item_code:
		bom_name = frappe.db.get_value(
			"BOM", {"item": item_code, "is_active": 1, "is_default": 1, "docstatus": 1}, "name"
		)
		result["bom"] = bom_name
		if bom_name:
			bom_items = frappe.get_all(
				"BOM Item", filters={"parent": bom_name}, fields=["item_code"], order_by="idx asc", limit_page_length=1
			)
			if bom_items:
				result["rm_item"] = bom_items[0].item_code

	if result["rm_item"] and warehouse:
		result["available_qty"] = frappe.db.get_value(
			"Bin", {"item_code": result["rm_item"], "warehouse": warehouse}, "actual_qty"
		) or 0

	return result

File: apil_custom/public/js/extrusion_log.js
// Live form calculations for Extrusion Log: Total Input, Die Running Time,
// Rec%, Output/Hr, stock availability indicator, and quick-nav buttons to
// the resulting Stock Entry / Stock Balance report.
//
// Migrated from a database-stored Client Script of the same name - identical
// behaviour, just living in the app instead of the database.

frappe.ui.form.on("Extrusion Log", {
	refresh: function (frm) {
		if (frm.doc.stock_entry) {
			frm.add_custom_button("View Stock Entry", function () {
				frappe.set_route("Form", "Stock Entry", frm.doc.stock_entry);
			});
		}
		if (frm.doc.included_in_shift_log) {
			frm.add_custom_button("View Shift Production Log", function () {
				frappe.set_route("Form", "Shift Production Log", frm.doc.included_in_shift_log);
			});
		}
		if (frm.doc.rm_item && frm.doc.source_warehouse) {
			frm.add_custom_button("View Billet Stock Balance", function () {
				frappe.route_options = { item_code: frm.doc.rm_item, warehouse: frm.doc.source_warehouse };
				frappe.set_route("query-report", "Stock Balance");
			});
		}
		show_stock_indicator(frm);
		set_cut_length_query(frm);
	},
	sec_no: function (frm) {
		fetch_availability(frm);
		set_cut_length_query(frm);
	},
	source_warehouse: function (frm) {
		fetch_availability(frm);
	},
	die_in: function (frm) {
		calc_running_time(frm);
	},
	die_out: function (frm) {
		calc_running_time(frm);
	},
	output: function (frm) {
		calc_rec(frm);
	},
});

frappe.ui.form.on("Extrusion Log Billet Charge", {
	billet_weight: function (frm) {
		calc_total_input(frm);
	},
	billet_charges_remove: function (frm) {
		calc_total_input(frm);
	},
});

function set_cut_length_query(frm) {
	// Only offer UOMs actually set up as alternates on this item (its Stock
	// UOM, e.g. standard "6.4M Length", plus any special-order lengths
	// added to its UOM table) - stops an operator picking an unrelated
	// length that has no conversion factor on this item at all.
	frm.set_query("cut_length", function () {
		return {
			query: "apil_custom.extrusion_log.query_item_uoms",
			filters: { item_code: frm.doc.sec_no },
		};
	});
}

function fetch_availability(frm) {
	if (!frm.doc.sec_no) return;
	frappe.call({
		method: "apil_custom.extrusion_log.get_extrusion_rm_availability",
		args: { item_code: frm.doc.sec_no, warehouse: frm.doc.source_warehouse },
		callback: function (r) {
			if (r.message) {
				frm.set_value("rm_item", r.message.rm_item);
				frm.set_value("available_stock", r.message.available_qty);
				frm.set_value("bom", r.message.bom);
				show_stock_indicator(frm);
				frm.refresh();
			}
		},
	});
}

function calc_total_input(frm) {
	let total = 0;
	(frm.doc.billet_charges || []).forEach(function (row) {
		total += flt(row.billet_weight);
	});
	frm.set_value("total_input", total);
	calc_rec(frm);
	show_stock_indicator(frm);
}

function calc_running_time(frm) {
	if (frm.doc.die_in && frm.doc.die_out) {
		let diff = moment.duration(frm.doc.die_out) - moment.duration(frm.doc.die_in);
		if (diff < 0) diff += 24 * 3600 * 1000;
		frm.set_value("die_running_time", diff / 1000);
		calc_output_per_hr(frm);
	}
}

function calc_rec(frm) {
	if (frm.doc.output && frm.doc.total_input) {
		frm.set_value("rec_percent", flt((frm.doc.output / frm.doc.total_input) * 100, 2));
	}
	calc_output_per_hr(frm);
}

function calc_output_per_hr(frm) {
	if (frm.doc.output && frm.doc.die_running_time) {
		let hours = frm.doc.die_running_time / 3600;
		if (hours > 0) frm.set_value("output_per_hr", flt(frm.doc.output / hours, 2));
	}
}

function show_stock_indicator(frm) {
	frm.dashboard.clear_headline();
	if (frm.doc.available_stock === undefined || frm.doc.available_stock === null) return;
	let needed = flt(frm.doc.total_input);
	let available = flt(frm.doc.available_stock);
	if (!needed) return;
	if (needed > available) {
		frm.dashboard.set_headline_alert(
			"Insufficient " + (frm.doc.rm_item || "billet") + " stock in " + frm.doc.source_warehouse +
				": available " + available + " kg, this log needs " + needed + " kg. Please refill stock before submitting.",
			"red"
		);
	} else {
		frm.dashboard.set_headline_alert(
			"Stock OK: " + available + " kg " + (frm.doc.rm_item || "billet") + " available (needs " + needed + " kg).",
			"green"
		);
	}
}
